src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,200 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+// This file describes AST nodes for authorization logic
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_H_
+#define SRC_IR_AUTH_LOGIC_AST_H_
+
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace raksha::ir::auth_logic {
+
+class Principal {
+ public:
+  Principal(std::string name) : name_(std::move(name)) {}
+
+ private:
+  std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum Sign { kNegated, kPositive };
+
+class Predicate {
+ public:
+  Predicate(std::string name, std::vector<std::string> args, Sign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        // TODO move surpresses an unused private field error about sign_
+        // for now, get rid of this eventually
+        sign_(std::move(sign)) {}
+
+ private:
+  std::string name_;
+  std::vector<std::string> args_;
+  Sign sign_;
+};
+
+// This represents an expression that applies either a predicate or a
+// "canActAs" to a principal, and it is a sub-term of FlatFact
+//    - pred_name(arg_1, ..., arg_n)           (a predicate)
+//    - principal_1 canActAs principal_2       (a canActAs)
+// To encapsulate which form this takes, it:
+//    - uses factory methods to make either form
+//    - uses a private std::variant member that takes either of these two forms
+//    - uses std::visit to do operations on the actual state
+class VerbPhrase {
+ public:
+  static std::unique_ptr<VerbPhrase> CreateVerbPhrase(Predicate pred) {
+    return std::unique_ptr<VerbPhrase>(new VerbPhrase(std::move(pred)));
+  }
+  static std::unique_ptr<VerbPhrase> CreateVerbPhrase(Principal right_prin) {
+    return std::unique_ptr<VerbPhrase>(new VerbPhrase(std::move(right_prin)));
+  }
+
+ private:
+  // The principal variant of the internal form means this represents
+  // "canActAs principal".
+  std::variant<Predicate, Principal> internal_form_;
+  VerbPhrase(std::variant<Predicate, Principal> internal_form)
+      : internal_form_(std::move(internal_form)) {}
+};
+
+// This the AST Node for FlatFact which takes on one of two forms:
+//   - <principal> <verb_phrase>  (* a principal applied to a verbphrase *)
+//   - <predicate>                (* just a predicate *)
+// A similar variant-based pattern is used for this as for VerbPhrase. Here an
+// inner struct is used for the first form (principal applied to verb_phrase).
+class FlatFact {
+ public:
+  static std::unique_ptr<FlatFact> CreateFlatFact(Principal prin,
+                                                  VerbPhrase verb_phrase) {
+    return std::unique_ptr<FlatFact>(
+        new FlatFact(VerbPhraseFlatFact(prin, verb_phrase)));
```

Comment:
You should probably `std::move` `prin` and `verb_phrase` here so that they have no live names when being passed into the `VerbPhraseFlatFact` constructor and can be copy elided.

Reply:
I thought the `std::move`s that the constructor for VerbPhraseFlatFact already does would be enough. In any case, I will add them.